exec-service-wasmer/src/wasmer_breakpoint.rs

```diff
@@ -0,0 +1,168 @@
+use std::mem;
+use std::sync::Mutex;
+
+use loupe::{MemoryUsage, MemoryUsageTracker};
+use wasmer::wasmparser::{Operator, Type as WpType, TypeOrFuncType as WpTypeOrFuncType};
+use wasmer::{
+    ExportIndex, FunctionMiddleware, GlobalInit, GlobalType, Instance, LocalFunctionIndex,
+    MiddlewareError, MiddlewareReaderState, ModuleMiddleware, Mutability, Type,
+};
+use wasmer_types::{GlobalIndex, ModuleInfo};
+
+const BREAKPOINT_VALUE: &str = "breakpoint_value";
+
+pub(crate) const BREAKPOINT_VALUE_NO_BREAKPOINT: u64 = 0;
+#[allow(dead_code)]
+pub(crate) const BREAKPOINT_VALUE_EXECUTION_FAILED: u64 = 1;
+pub(crate) const BREAKPOINT_VALUE_OUT_OF_GAS: u64 = 4;
+#[allow(dead_code)]
+pub(crate) const BREAKPOINT_VALUE_MEMORY_LIMIT: u64 = 5;
+
+#[derive(Clone, Debug, MemoryUsage)]
+struct BreakpointGlobalIndex(GlobalIndex);
+
+impl BreakpointGlobalIndex {
+    fn breakpoint_value(&self) -> GlobalIndex {
+        self.0
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Breakpoint {
+    global_indexes: Mutex<Option<BreakpointGlobalIndex>>,
+}
+
+impl Breakpoint {
+    pub(crate) fn new() -> Self {
+        Self {
+            global_indexes: Mutex::new(None),
+        }
+    }
+
+    pub(crate) fn insert_breakpoint<'b>(&self, value: u64) -> Vec<Operator<'b>> {
+        vec![
+            Operator::If {
+                ty: WpTypeOrFuncType::Type(WpType::EmptyBlockType),
+            },
+            Operator::I64Const {
+                value: value as i64,
+            },
+            Operator::GlobalSet {
+                global_index: self.get_breakpoint_value(),
+            },
+            Operator::End,
+        ]
+    }
+
+    fn get_breakpoint_value(&self) -> u32 {
+        self.global_indexes
+            .lock()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .breakpoint_value()
+            .as_u32()
+    }
+}
+
+unsafe impl Send for Breakpoint {}
+unsafe impl Sync for Breakpoint {}
+
+impl ModuleMiddleware for Breakpoint {
+    fn generate_function_middleware(
+        &self,
+        _local_function_index: LocalFunctionIndex,
+    ) -> Box<dyn FunctionMiddleware> {
+        Box::new(FunctionBreakpoint {
+            global_indexes: self.global_indexes.lock().unwrap().clone().unwrap(),
+        })
+    }
+
+    fn transform_module_info(&self, module_info: &mut ModuleInfo) {
+        let mut global_indexes = self.global_indexes.lock().unwrap();
+
+        if global_indexes.is_some() {
+            panic!("Breakpoint::transform_module_info: Attempting to use a `Breakpoint` middleware from multiple modules.");
```

Comment:
What does this condition actually refer to?

Do `modules` here mean live WASM instances? `Singlepass` would never reuse the same middleware instance for multiple WASM instances. Right?

Reply:
I think we can remove this condition, you are right.